### PR TITLE
Rewrite macOS IOHIDManager communication and state machine

### PR DESCRIPTION
**Pull request by ttaubert**
r? @jcjones 

Looking forward to your feedback!

This seems very stable, I can't make a custom Firefox panic when plugging/unplugging multiple times. Re-plugged devices are properly registered, and work for registration and signing. With Chrome open, I also can't reproduce bug 1398268 anymore.

### CODE REVIEW

File: src/macos/mod.rs [hunk 1]
```diff
@@ -47,56 +41,43 @@ impl PlatformManager {
...
+            let dev = &mut Device::new(device_ref, rx);
+
+            // Try initializing it.
+            if !u2f_init_device(dev) {
+                return;
+            }
+
+            // Iterate the exlude list and see if there are any matches.
+            // Abort the state machine if we found a valid key handle.
+            if key_handles.iter().any(|key_handle| {
```
Comment (jcjones) [inline src/macos/mod.rs]: Nit: feels like you should use the same chaining style here as around line 105... either's fine, but ... one. :)

File: src/macos/monitor.rs [hunk 2]
```diff
@@ -2,115 +2,175 @@
...
-        let (tx, rx) = channel();
+impl<F> Monitor<F>
+where
+    F: Fn(IOHIDDeviceRef, Receiver<Vec<u8>>, &Fn() -> bool) + Sync + 'static,
+{
+    pub fn new(new_device_cb: F) -> Self {
+        let manager = unsafe { IOHIDManagerCreate(kCFAllocatorDefault, kIOHIDManagerOptionNone) };
+
+        // Match FIDO devices only.
+        let _matcher = IOHIDDeviceMatcher::new();
```
Comment (jcjones) [inline src/macos/monitor.rs]: btw, I'm super impressed that you figured out how to make this work. I fought it ... too long.

File: src/macos/monitor.rs [hunk 3]
```diff
@@ -2,115 +2,175 @@
...
+
+            IOHIDManagerScheduleWithRunLoop(
+                self.manager,
+                CFRunLoopGetCurrent(),
+                kCFRunLoopDefaultMode,
+            );
+
+            let rv = IOHIDManagerOpen(self.manager, kIOHIDManagerOptionNone);
+            if rv != 0 {
+                return Err(io_err("Couldn't open HID Manager"));
```
Comment (jcjones) [inline src/macos/monitor.rs]: Nit: Probably good to put the `rv` into the message, but *shrug*

File: src/macos/monitor.rs (hunk 2 above)
Comment (ttaubert) [reply src/macos/monitor.rs]: It took... at lot of time :)

File: src/macos/mod.rs (hunk 1 above)
Comment (ttaubert) [reply src/macos/mod.rs]: They're not quite the same. For `register()` we want to bail out as soon as we match one of the entries in the exclusion list. In `sign()` we want to actually get a list of all matching tokens and try to sign with those one by one.

I'll leave this as is and land it. If you disagree, let's open an issue and I'll be happy to address that in a follow-up :)

File: src/macos/monitor.rs (hunk 3 above)
Comment (ttaubert) [reply src/macos/monitor.rs]: Good idea, done.

**Pull request closed without merging**